### ISSUE TITLE
__mul function error

### ISSUE THREAD

**Issue by yongfei25** [twitter/torch-autograd#54]
Hi,

I have this expression below in a function within `autograd(...)`

```lua
-- x is a table of Nodes, { Node, Node, ... }
local xt = torch.view(x[1], 1, inputFeatures)

-- line 77
dots = torch.cat(xt,hp,2) * p.W + torch.expand(p.b, 1, 2*hiddenFeatures)
```

and keep getting

```lua
/home/yongfei/torch/install/bin/luajit: bad argument #2 to '?' (number expected, got table)
stack traceback:
	[C]: at 0x7f3b08895ac0
	[C]: in function '__mul'
	/home/yongfei/codes/dmn-1/model/SimpleGRU.lua:77: in function 'episodeGRU'
	/home/yongfei/codes/dmn-1/model/DMN.lua:184: in function 'dmn'
	train.lua:112: in function 'fun'
	...rch/install/share/lua/5.1/autograd/direct/DirectTape.lua:19: in function 'funOnly'
	...rch/install/share/lua/5.1/autograd/direct/DirectTape.lua:114: in function 'dDmn'
	train.lua:122: in main chunk
	[C]: in function 'dofile'
	...gfei/torch/install/lib/luarocks/rocks/trepl/scm-1/bin/th:133: in main chunk
	[C]: at 0x00406670
```

Here's the values in the expression

```lua
xt.value ~= nil	
> false	
xt:type()	
> torch.DoubleTensor	

hp.value ~= nil	
> false	
hp:type()	
> torch.DoubleTensor	

p.W.value ~= nil	
> true	
p.W.value:type()	
> torch.DoubleTensor	

(torch.cat(xt,hp,2).value) ~= nil	
> false	
(torch.cat(xt,hp,2)):type()	
> torch.DoubleTensor	
```

**Comment by alexbw** [comment]
Hey @yongfei25, could you give us a full runnable example? That way we can insert it as a test case and work on it from there.

**Comment by yongfei25** [comment]
@alexbw I think most probably is because I have nested autograd function, like this

```lua
d(
  function(params, params2)
    d(function(params2)  
      -- ...
      return loss
    end)
  end
)
```

Is the structure above allowed? Second params are not auto-differentiated in the outer function.

**Comment by yongfei25** [comment]
You may clone the codes here: git@github.com:yongfei25/dmn-1.git,  start the example with `th train.lua`

Here's the inner autograd function I mentioned: https://github.com/yongfei25/dmn-1/blob/master/model/DMN.lua#L128

**Comment by alexbw** [comment]
I think nested grads are almost supported, but not entirely cc @luketwitter

**Comment by yongfei25** [comment]
```lua
require 'torch'
local d = require 'autograd'
local model = require 'autograd.model'

torch.manualSeed(123)

local params = {
  W = torch.Tensor(1, 100):uniform(-0.1,0.1)
}

local lstm2, params2 = model.RecurrentLSTMNetwork({
  inputFeatures = 50,
  hiddenFeatures = 100,
  outputType = 'last'
})

local lsm = d.nn.LogSoftMax()
local lossf = d.nn.ClassNLLCriterion()

local inputs = torch.Tensor(10, 50):uniform(-0.1, 0.1)
local y = torch.Tensor(1, 1):uniform(-0.1, 0.1)

for i,v in pairs(params2[1]) do
  params2[1][i] = params2[1][i]:uniform(-0.1, 0.1)
end

local f2 = function(params2, x, y)
  local h1, newState1 = lstm2(params2, x)
  local h2f = torch.view(h1, 1, 100)
  local yhat = lsm(h2f)
  local loss = lossf(yhat, torch.Tensor{1})

  return loss
end

local df2 = d(f2)

local f = function(params, params2, x, y)
  local grads, loss = df2(params2, x, y)

  local dot = x * params.W -- error here
  return loss
end


local df = d(f)

df(params, params2, inputs, y)
```

**Comment by alexbw** [comment]
So we're close to enabling this, which ends up being a dupe of #4. 
The trick is that we need to be able to take gradients of gradients (since you have nested them). There are a few functions whose gradients use functions for which we haven't defined gradients. Once we have those written, this should work.

**Comment by alexbw** [comment]
@yongfei25 Can you try again? There should be nested grad functionality now. Also, it would help us iterate quickly if you provided a runnable test instead of pseudocode.

**Comment by yongfei25** [comment]
Hey @alexbw , thanks for updating.

I'm still having some weird error, gradients returned by nested grad function are empty. Below is a piece of runnable example.

```lua
require 'torch'
d = require 'autograd'
model = require 'autograd.model'

torch.manualSeed(123)

local opt = { nWords = 5, wordSize = 50, hiddens = 200, nClass1 = 10, nClass2 = 4}

local lstm,lParams = model.RecurrentLSTMNetwork({
   inputFeatures = opt.wordSize,
   hiddenFeatures = opt.hiddens,
   outputType = 'last',
})

table.insert(lParams, {W2 = torch.Tensor(opt.hiddens, opt.nClass2)})

-- init params
for km, v in pairs(lParams) do
  for k, v2 in pairs(lParams[km]) do
    lParams[km][k]:uniform(-0.1,0.1)
  end
end

-- vars for inner grad
local inputs = torch.Tensor(opt.nWords, opt.wordSize):uniform(-0.1, 0.1)
local sigmoid = d.nn.Sigmoid()
local lossf2 = d.nn.BCECriterion()
local prevState = torch.Tensor(1, opt.hiddens):uniform(-0.1, 0.1)

local df2 = d(function(lParams, x, y)
  local state = lstm(lParams, x, prevState)
  local yhat = sigmoid(state * lParams[2].W2)

  local loss = lossf2(yhat, y)
  return loss, state
end)

-- vars for outer grad
local params = {
  W = torch.Tensor(200, opt.nClass1):uniform(-0.1, 0.1),
  b = torch.Tensor(1, opt.nClass1):uniform(-0.1, 0.1)
}
local lsm = d.nn.LogSoftMax()
local lossf = d.nn.ClassNLLCriterion()

local df = d(function(params, lParams, x, y)

  local stop = false
  local stopLoss = 0.01
  local ly = torch.Tensor({1,0,0,1})
  local count = 0
  local chosenState = torch.view(torch.Tensor(opt.hiddens):uniform(-0.1, 0.1), 1,opt.hiddens)

  -- nested training with autograd
  repeat
    count = count + 1
    local grads, loss, chosenState = df2(lParams, x, ly)

    if count % 50 == 0 then print('inner grad: ' .. loss) end

    if loss.raw < stopLoss then
      stop = true
      print('inner grad: ' .. loss)
    else
      -- update params
      for km, v in pairs(lParams) do
        for k, v2 in pairs(lParams[km]) do
          lParams[km][k]:add(-0.1, grads[km][k])
        end
      end
    end
  until stop

  local yhat = lsm(chosenState * params.W + params.b)
  local loss = lossf(yhat, y)
  return loss
end)

local y = torch.Tensor({6})
local grads, loss = df(params, lParams, inputs, y)
print(loss)
```

Error
```
/home/yongfei/torch/install/bin/luajit: train-nested.lua:69: invalid arguments: table number nil 
expected arguments: *DoubleTensor* [DoubleTensor] double | *DoubleTensor* [DoubleTensor] [double] DoubleTensor
stack traceback:
	[C]: in function 'add'
	train-nested.lua:69: in function 'fn'
	...install/share/lua/5.1/autograd/runtime/codegen/Graph.lua:281: in function 'protectedFn'
	...install/share/lua/5.1/autograd/runtime/codegen/Graph.lua:309: in function 'record'
	.../install/share/lua/5.1/autograd/runtime/codegen/init.lua:44: in function 'generateFn'
	.../install/share/lua/5.1/autograd/runtime/codegen/init.lua:66: in function 'df'
	train-nested.lua:81: in main chunk
	[C]: in function 'dofile'
	...gfei/torch/install/lib/luarocks/rocks/trepl/scm-1/bin/th:145: in main chunk
	[C]: at 0x00406670
```

**Comment by yongfei25** [comment]
I checked my code after some times, realized it was  due to a bug in my code. Closing .

**Comment by alexbw** [comment]
Cool. Did you end up using nested gradients?

**Comment by yongfei25** [comment]
I didn't. I was stucked and then I moved to implement something else :\ . Sorry haven't really test it yet. And thanks for the update :)
